toolchain/check/class.cpp

```diff
@@ -0,0 +1,269 @@
+       .virtual_functions_id = context.inst_blocks().Add(vtable)});
+}
+
+// Checks that the specified finished class definition is valid and builds and
+// returns a corresponding complete type witness instruction.
+static auto CheckCompleteClassType(
+    Context& context, Parse::NodeId node_id, SemIR::ClassId class_id,
+    llvm::ArrayRef<SemIR::InstId> field_decls,
+    llvm::ArrayRef<SemIR::InstId> vtable_contents,
+    llvm::ArrayRef<SemIR::InstId> inst_block) -> SemIR::InstId {
+  auto& class_info = context.classes().Get(class_id);
+  if (class_info.adapt_id.has_value()) {
+    return CheckCompleteAdapterClassType(context, node_id, class_id,
+                                         field_decls, inst_block);
+  }
+
+  bool defining_vptr = class_info.is_dynamic;
+  auto base_type_id =
+      class_info.GetBaseType(context.sem_ir(), SemIR::SpecificId::None);
+  SemIR::Class* base_class_info = nullptr;
+  if (base_type_id.has_value()) {
+    // TODO: If the base class is template dependent, we will need to decide
+    // whether to add a vptr as part of instantiation.
+    base_class_info = TryGetAsClass(context, base_type_id);
+    if (base_class_info && base_class_info->is_dynamic) {
+      defining_vptr = false;
+    }
+  }
+
+  llvm::SmallVector<SemIR::StructTypeField> struct_type_fields;
+  struct_type_fields.reserve(defining_vptr + class_info.base_id.has_value() +
+                             field_decls.size());
+  if (defining_vptr) {
+    struct_type_fields.push_back(
+        {.name_id = SemIR::NameId::Vptr,
+         .type_id = GetPointerType(
+             context,
+             GetSingletonType(context, SemIR::VtableType::SingletonInstId))});
+  }
+  if (base_type_id.has_value()) {
+    auto base_decl = context.insts().GetAs<SemIR::BaseDecl>(class_info.base_id);
+    base_decl.index =
+        SemIR::ElementIndex{static_cast<int>(struct_type_fields.size())};
+    ReplaceInstPreservingConstantValue(context, class_info.base_id, base_decl);
+    struct_type_fields.push_back(
+        {.name_id = SemIR::NameId::Base, .type_id = base_type_id});
+  }
+
+  if (class_info.is_dynamic) {
+    class_info.vtable_id = BuildVtable(
+        context, node_id,
+        defining_vptr ? SemIR::InstId::None : base_class_info->vtable_id,
+        vtable_contents);
+  }
+
+  return AddInst<SemIR::CompleteTypeWitness>(
+      context, node_id,
+      {.type_id =
+           GetSingletonType(context, SemIR::WitnessType::SingletonInstId),
+       .object_repr_id = GetStructType(
+           context,
+           AddStructTypeFields(context, struct_type_fields, field_decls))});
+}
+
+auto ComputeClassObjectRepresentation(
+    Context& context, Parse::NodeId node_id, SemIR::ClassId class_id,
+    llvm::ArrayRef<SemIR::InstId> field_decls,
+    llvm::ArrayRef<SemIR::InstId> vtable_contents,
+    llvm::ArrayRef<SemIR::InstId> inst_block) -> void {
+  // The class type is now fully defined. Compute its object representation.
```

Comment:
This comment may have made sense in the code you're moving, but now it seems confusing -- what ensured this code is fully defined, other than the function comment that said so?

I see it's remained in handle_class.cpp, which might be where this really belongs -- so removing this would clean up a confusing duplication.

```suggestion
```

Reply:
Done.